articles/ros2_security_contexts.md

```diff
@@ -0,0 +1,234 @@
+---
+layout: default
+title: ROS 2 Security Contexts
+permalink: articles/ros2_access_control_policies.html
+abstract:
+  This article specifies the integration between security and contexts.
+author:  >
+  [Ruffin White](https://github.com/ruffsl),
+  [Mikael Arguedas](https://github.com/mikaelarguedas)
+
+published: false
+categories: Security
+---
+
+{:toc}
+
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+
+TODO: Some concise overview introduction here
+
+## Concepts
+
+Before detailing the SROS 2 integration of the contexts, the following concepts are introduced.
+
+### Namespaces
+
+Namespaces are a fundamental design pattern in ROS and are wildly used to organize and differentiate many types of resource as to be uniquely identifiable; i.e. for topics, services, actions, and node names. As such, the concept of namespaceing is well know and understood by current users, as well as strongly supported with the existing tooling. Namespaces are often configurable at runtime via command line arguments or statically/programmatically via launch file decelerations.
+
+Previously, the Fully Qualified Namespace (FQN) of a node was used directly by a selected security directory lookup strategy to load the necessary key material. However, with the advent of contexts, such a direct mapping of FQN to security artifacts may no longer suffice.
+
+### Contexts
+
+With the advent of ROS 2, multiple nodes may now be composed into one process for improved performance. Previously however, each node would retain it's one to one mapping to a septate middleware participant. Given the non-negligible overhead incurred of multiple participants per process, a concept of contexts was introduced. Contexts permit a many-to-one mapping of nodes to participant by grouping many nodes per context, and one context per participant.
+
+Currently, DDS participants can only utilise a single security identity; consequently the access control permissions applicable to every node mapped to a given context must be consolidated and combined into a single set of security artifacts. As such, additional tooling and extensions to SROS 2 are necessary to support this new paradigm.
+
+
+## Keystore
+
+With the additional structure of contexts, it’s perhaps best to take the opportunity to restructure the keystore layout as well. Rather than a flat directory of namespaced node security directories, we can push all such security directories down into a designated `contexts` sub-folder. Similarly, private and public keystore materials can also be pushed down into their own respective sub-folders within the root keystore directory. This is reminiscent of the pattern used earlier Keymint [1].
```

Comment:
> What I got from offline discussion was that this redesign was about the final identity to be the context FQN instead of the node FQN. so /context_namespace/context_name instead of /node_namespace/node_name

Correct, the node's name would not be used to resolve the context, allowing for nodes within the same namespace to easily be merged into a common context, or the default root context.

> The default context_namespace being / and the default context_name being . so that a participant using a default context would be finding artifacts directly at the keystore's root.

The change I added was that root context points to the root of the contexts folder, not the root of the keystore folder, just to keep the root of the keystore file system hygienic and organized. Feel free suggest a rephrasing to clarify this  this point.